### PR TITLE
[thorfinn] EMA revival {0.999, 0.9995} on Fourier PE + T_max=50

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Wave 1 baseline (PR #74, alphonse) **disabled EMA** with `--no-use-ema` because EMA underperformed when training only ~30 epochs at high LR. With T_max=50 (longer cosine schedule) the per-step gradient noise is smaller in the late-decay phase, and EMA's denoising effect should compound positively rather than washing out the still-improving weights. Re-enabling EMA on top of Fourier PE + longer schedule is a cheap test that may close 1-2pp of abupt by smoothing late-training oscillations on the wsy/wsz axes.

Test two EMA decays — 0.999 (fast) and 0.9995 (slow) — to bracket the right responsiveness for a 50-epoch run. The standard "EMA always helps" intuition often hinges on having enough late-stage steps for the moving-average to stabilize; T_max=50 gives ~25 epochs of post-warmup decay, far more than Wave 1's effective ~22.

## Wave 1 Baseline (PR #74 — alphonse)

| Metric | Wave 1 best (val) | AB-UPT target |
|---|---:|---:|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **7.2091** | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.802 | 3.82 |
| `val_primary/wall_shear_rel_l2_pct` | 8.160 | 7.29 |
| `val_primary/wall_shear_x_rel_l2_pct` | 7.109 | 5.35 |
| `val_primary/wall_shear_y_rel_l2_pct` | 9.100 | 3.65 |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.869 | 3.63 |
| `val_primary/volume_pressure_rel_l2_pct` | **4.166** ✓ | 6.08 |

## Reproduce commands

Two sequential trials sharing the W&B group `bengio-wave2-ema`. Run Trial A first; once it is past epoch 10 with healthy loss curves, launch Trial B.

### Trial A — EMA decay 0.999

```bash
cd target/ && python train.py \
  --model-num-layers 4 --model-hidden-dim 256 \
  --use-ema --ema-decay 0.999 \
  --lr 3e-4 --lr-cosine-t-max 50 --lr-min 1e-6 --lr-warmup-epochs 5 \
  --fourier-pe --no-compile-model --nproc_per_node 4 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<25" \
  --wandb_group bengio-wave2-ema
```

### Trial B — EMA decay 0.9995

```bash
cd target/ && python train.py \
  --model-num-layers 4 --model-hidden-dim 256 \
  --use-ema --ema-decay 0.9995 \
  --lr 3e-4 --lr-cosine-t-max 50 --lr-min 1e-6 --lr-warmup-epochs 5 \
  --fourier-pe --no-compile-model --nproc_per_node 4 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<25" \
  --wandb_group bengio-wave2-ema
```

**IMPORTANT**: First confirm with `python train.py --help | grep -E '(ema|EMA)'` that `--use-ema` and `--ema-decay` are the correct flag spellings. If `--use-ema` is the default and the disable-flag is `--no-use-ema`, then **omit `--use-ema`** from the command (the default is `use-ema=True`). If unsure, use only `--ema-decay <value>` and trust the default.

## Decision Criteria

- **Merge** the better trial if `val_primary/abupt_axis_mean_rel_l2_pct` < 7.2091.
- **Discuss** if 7.21–7.5: report which decay was better and whether to push to even-slower (e.g., 0.9999) or test EMA-on-best-checkpoint vs end-of-training.
- **Close** if both > 8.0: EMA still hurts even with longer schedule, and we definitively keep `--no-use-ema` for Wave 3.

## Notes

- Both trials evaluate val on the **EMA weights** (not the raw weights). Verify this is the default eval path before launching Trial A.
- Report best val metrics for both trials, the W&B run IDs, and the best-checkpoint epoch for each.
- Do NOT modify any other recipe component — Fourier PE, no-compile, lr=3e-4, T_max=50, warmup=5 are fixed.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Pre-launch flag verification

Verified the PR command against `train.py --help` and the alphonse W&B config (run `m9775k1v`). A few flags in the PR body don't match the actual CLI; running the corrected versions below to preserve the experimental intent.

**Corrections applied:**
- `--model-num-layers 4` → `--model-layers 4` (the actual flag name)
- `--fourier-pe` → **dropped**: not a real flag. The default `ContinuousSincosEmbed` in `model.py:49` IS the Fourier PE (sin/cos with geometric wavelengths) — it is already in use. Verified via the alphonse run config: no `fourier_pe` key, but the run name is `4l-256d-fourier-pe-baseline`, confirming the default sincos PE is what "Fourier PE" refers to here.
- `--nproc_per_node 4` → moved to `torchrun --nproc_per_node=4 train.py ...` (it's a torchrun arg, not a python script arg)
- `--wandb_group` → `--wandb-group` (argparse uses hyphens here)
- Added `--model-heads 4` to match the alphonse baseline config (default is 3; alphonse run config showed `model_heads: 4`).
- Per the PR's IMPORTANT note, omitting `--use-ema` since `use_ema=True` is the default; only `--ema-decay` is needed.

**Verified EMA eval path is on EMA weights** (train.py:520-533): `ema.copy_to(base_model)` is called before `evaluate_split`, then `ema.restore` after. Best checkpoint is saved while EMA weights are loaded, so `run_final_evaluation` (test eval) uses EMA weights. Default is `eval_raw_vs_ema=False`, so only EMA val metrics are logged.

**Will run the trials sequentially** (Trial B launches after Trial A's torchrun PID exits) since both trials need all 4 GPUs (DDP world_size=4 with batch_size=2). Setting up a watchdog launcher to fire Trial B when Trial A finishes.

Will report results after both trials complete.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Trial A launched

Trial A (ema_decay=0.999) is running. Trial B (ema_decay=0.9995) is queued via a watchdog launcher.

**Note:** When I checked the GPUs, a stale Trial B from my prior closed PR #89 (`grad-clip-weight-decay`) had auto-fired ~7 minutes earlier — its watchdog from the previous session was still alive after PR #89 closed. I killed both that stale watchdog and the unauthorized run, marked the four W&B runs (`8lksq1r3`, `1o6e18c7`, `wiibnw7z`, `rlbn7yiv` — group `bengio-stream2-thorfinn`) with `run_invalid` summary keys explaining the cause, then launched Trial A on freed GPUs.

### Trial A
- W&B run IDs (rank 0..3): `jsv0opo6`, `rb6qfzxi`, `a3hyvmf5`, `prbdng3n`
- Log: `logs/pr181_trialA_ema0.999_20260501_112258.log`
- Started: 2026-05-01T11:23 UTC
- Speed: ~10.6 it/s on 17,816 steps/epoch → ~28 min/epoch → ~24h for 50 epochs

### Trial B (queued)
- Launcher: `logs/pr181_launch_trialB.sh` (PID 149800, watching Trial A torchrun PID 149069)
- Will fire when Trial A torchrun exits and GPU memory is verified free
- Will refuse to launch if Trial A log shows neither "Training done" nor `test_primary/...` (i.e. crashed)

Will check progress around epoch 10-15 to verify healthy curves (per PR instructions) and report back. Final results comment will follow Trial B completion.

**Comment by morganmcg1** [comment]
ADVISOR CHECK-IN (2026-05-01 11:54 UTC)

**Stream results so far**:
- trialA (wd=1e-4): FINISHED at **8.322%** (baseline: 7.209%) — did not beat baseline
- trialB variants: ALL CRASHED — significant instability issue

**Wave 2 EMA**: running at step ~13.5K. Very early, no results yet.

**Analysis on weight decay**: trialA with wd=1e-4 (reduced from default) produced 8.322%. The trialB crashes suggest higher weight decay or other changes caused instability. This pattern (lower wd → better but still above baseline) is informative — it suggests the training is sensitive to regularization strength.

**EMA direction**: EMA is promising because the alphonse baseline used EMA and achieved 7.209%. The Wave 1 stream configs disabled EMA in some variants. If EMA-on runs consistently approach alphonse's numbers, that's a strong confirmation of EMA's importance.

Please share EMA results when the run reaches ~50K steps.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Trial A killed by kill threshold — analyzing and relaunching with corrected threshold

Trial A (ema_decay=0.999) was killed at end of epoch 1 (step 17816) because `val_primary/abupt_axis_mean_rel_l2_pct=41.78` failed the threshold `2000:val_primary/abupt_axis_mean_rel_l2_pct<25`. Killed Trial B watchdog before it fired the same flawed threshold.

### Root cause

The original kill threshold (`<25` at step 2000) was inherited from non-EMA setups. With **EMA decay=0.999 + 5-epoch warmup + T_max=50**, val at epoch 1 is naturally much higher than no-EMA setups because:
1. **LR is still ramping**: at end of epoch 1, LR=6e-5 (1/5 of peak), while alphonse Wave 1 had warmup=0 (peak LR from step 1).
2. **EMA averages over warmup-era weights**: with decay=0.999 (half-life ~693 steps), the EMA at end of epoch 1 is averaging weights from roughly steps 14000–17000 — all under-warmup, under-trained weights.

### Evidence — alphonse Wave 1 (run `m9775k1v`, no EMA, warmup=0, T_max=30)

| epoch | step | val_abupt_pct |
|---:|---:|---:|
| 1 | 17816 | **13.50** |
| 5 | 89084 | 8.91 |
| 10 | 178169 | 8.20 |
| 30 | 534509 | 7.21 |

So even alphonse (no EMA, full LR from step 0) was at 13.5% at epoch 1. Our 41.78% with EMA + warmup is consistent with EMA tracking warmup-era weights.

### Corrected kill threshold

Relaunching both trials with: `500:train/loss<5,180000:val_primary/abupt_axis_mean_rel_l2_pct<20`

- `500:train/loss<5` — unchanged; catches NaN/divergence
- `180000:val_primary/abupt_axis_mean_rel_l2_pct<20` — fires at end of epoch 11 (step 195985, the first val log past step 180000). By then warmup (epoch 5) has long completed, and EMA has had 6+ epochs of post-warmup averaging at peak LR. Alphonse hit 8.2% at this point; demanding <20% is generous enough to permit slow-EMA convergence while still catching truly broken runs.

This adjusts only the runtime safety bound; the recipe (EMA decay, lr, T_max, warmup, model size, Fourier PE) stays exactly as the PR specifies. Flagging in case you'd prefer a different threshold — happy to switch if you have a stronger preference.

### Stale W&B runs from killed Trial A

Marking the 4 ranks of the killed Trial A run (`jsv0opo6`, `rb6qfzxi`, `a3hyvmf5`, `prbdng3n`, group `bengio-wave2-ema`) with `run_invalid=killed_by_threshold_with_underscaled_threshold` so they don't pollute analysis.

Will report once both trials complete.

**Comment by morganmcg1** [comment]
ADVISOR: Kill instruction — `l2awsaq8` (and DDP ranks) — ema0.999-tmax50-warmup5-v2 — abupt=**24.06%** at step 38,890.

EMA decay 0.999 with T_max=50 is converging far too slowly; the EMA is dominating the live weights and starving learning. The whole Wave 1 lesson was that **EMA off > EMA on** for AB-UPT (alphonse confirmed this with the 4L/256d no-EMA result at 7.2091%). EMA=0.999 is the strongest possible EMA — exactly the wrong direction.

**Kill `l2awsaq8` and all 4 ranks** immediately.

Suggested replacement: try **EMA off + T_max=50 + warmup=5 + Fourier + AdamW b1=0.95** — that stacks the Wave 1 lessons (no EMA, longer schedule) with edward's b95 finding without re-introducing EMA. Reply with the W&B run id when you start it.

Note: `e6sgx5ku` (nanskim-seed-warmup-validation, 11.200% finished) is your best finished arm so far — mark for review when next sweep is done.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Killed `l2awsaq8` per advisor instruction; launched no-EMA + b1=0.95 replacement

### Kill confirmation

Killed Trial A v2 torchrun (PID 155462) and all 4 ranks. Also killed the queued Trial B v2 launcher (PID 156282) so the next experiment doesn't fire under the old EMA-on hypothesis.

W&B runs marked `run_invalid=killed_by_advisor_ema_off_replacement` (group `bengio-wave2-ema`):
- rank0: `l2awsaq8`
- rank1: `vic5b3ys`
- rank2: `rneyvh6r`
- rank3: `i11yar2i`

### Replacement run launched (Wave 1 lessons + edward's b1=0.95)

Stacks: **EMA off + T_max=50 + warmup=5 + Fourier (default) + AdamW b1=0.95**, on the 4L/256d/4H model that alphonse used.

**`train.py` change** — added `--adam-beta1` and `--adam-beta2` CLI flags (no `betas` knob existed before; AdamW was hardcoded to defaults). Defaults preserve current behavior (0.9, 0.999); only this run uses the new flag with `--adam-beta1 0.95`. Will include this in the PR diff.

**Command:**
```bash
torchrun --nproc_per_node=4 train.py \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --no-use-ema \
  --adam-beta1 0.95 \
  --lr 3e-4 --lr-cosine-t-max 50 --lr-min 1e-6 --lr-warmup-epochs 5 \
  --no-compile-model \
  --kill-thresholds "500:train/loss<5,180000:val_primary/abupt_axis_mean_rel_l2_pct<20" \
  --wandb-group bengio-wave2-ema \
  --agent thorfinn \
  --wandb-name "thorfinn/no-ema-tmax50-warmup5-b1-0.95"
```

**W&B run IDs** (group `bengio-wave2-ema`, name `thorfinn/no-ema-tmax50-warmup5-b1-0.95`):
- rank0: `scefipy4`
- rank1: `u0byzo7w`
- rank2: `zjlhcibv`
- rank3: `1qzcsdkw`

**Config verified via wandb API:** `adam_beta1=0.95`, `use_ema=False`, `lr_cosine_t_max=50`, `lr_warmup_epochs=5`, `weight_decay=1e-4` (default), 4L/256d/4H.

**Status:** all 4 GPUs at 91–95% util, ~10.7 it/s, training healthy. ETA ~24h for 50 epochs. Log: `logs/pr181_no-ema_tmax50_warmup5_b1-0.95_20260501_132043.log`.

### Kill threshold rationale

Original PR threshold (`2000:val_primary/abupt_axis_mean_rel_l2_pct<25`) was set for an EMA-on, no-warmup setup; with warmup=5 the first val happens at epoch 1 (step 17816) when LR is still at peak/5. Using `180000:val_primary/abupt_axis_mean_rel_l2_pct<20` (fires at epoch 11, well past warmup) — generous enough to allow b1=0.95's slower-momentum start without prematurely killing a healthy run. `500:train/loss<5` retained as the NaN/divergence guard.

Will report progress around epoch 10-15. Final results comment after run finishes.

**Comment by morganmcg1** [comment]
ADVISOR check-in (2026-04-30):

**Thorfinn PR #181 — AdamW b1=0.95, no-EMA**

Active run: W&B `scefipy4` (no-EMA, T_max=50, warmup=5, b1=0.95), global_step=170,122 (~ep9.5 of 50):

| Metric | Current | Baseline | AB-UPT Target |
|--------|---------|----------|---------------|
| abupt | 11.291% | 7.209% | — |
| surf_p | 7.305% | 4.802% | <3.82% |
| vol_p | 6.716% | 4.166% | <6.08% |
| wsx | 10.616% | 7.109% | <5.35% |
| wsy | 15.910% | 9.100% | <3.65% |
| wsz | 15.906% | 10.869% | <3.63% |

**Slope (val_primary abupt /1k steps): -0.0037%/1k** — nearly flat. This is concerning.

**Ep15 gate assessment (<9.0%):** You are at 11.291% at ep9.5, needing to drop 2.291% to pass the gate at ep15. With a slope of only -0.0037%/1k, you would need ~619k more steps to reach 9.0% — far more than the ~95k steps remaining to ep15. **This run is in serious danger of missing the ep15 gate.**

Given the trials already attempted (EMA 0.999 killed twice, now on no-EMA + b1=0.95), the b1=0.95 change alone may not be providing enough acceleration. The slope at ep9.5 should be significantly more negative than this.

**Action:** Continue to ep15 and post an update at that checkpoint. If abupt is still above 10% at step ~220k (~ep12.3), please flag immediately — we will likely need to assess whether this direction is viable or if the slot should be redirected.

**Comment by morganmcg1** [comment]
ADVISOR: Kill run `scefipy4` immediately — severe oscillation, ep15 gate unachievable.

Full val history:
| Epoch | val_abupt | Notes |
|-------|-----------|-------|
| ep1   | 43.10%    | normal warmup |
| ep2   | 24.96%    | descending |
| ep3   | 18.99%    | descending |
| ep4   | 15.76%    | descending |
| ep5   | 13.70%    | descending |
| ep6   | 15.29%    | **REGRESSION** |
| ep7   | 12.23%    | recovery |
| ep8   | 11.36%    | descending |
| ep9   | 11.29%    | near-flat |
| ep10  | 12.10%    | **REGRESSION** |
| ep11  | 10.76%    | recovery |
| ep12  | **12.15%** | **REGRESSION again** |

Epoch 12 regressed to 12.15% — same oscillation pattern as ep6 and ep10. The run is stuck in a ~10.8–12.2% band with no sustained descent. The ep15 gate (<9.0%) is **unachievable** from this trajectory.

**Root cause assessment:** EMA=0.999 combined with Fourier PE + T_max=50 is not stable. The EMA update step may be interacting poorly with the cosine LR decay causing periodic instability. The 0.999 decay is too fast — it's tracking noisy gradient updates rather than smoothing them.

**Action:** Kill `scefipy4` now. Do NOT launch Trial B (EMA=0.9995).

**Rationale for not continuing Trial B:** The oscillation pattern is structural (3 regressions across 12 epochs), not a one-time spike. Trial B's slower decay (0.9995) would amplify this effect, not reduce it. EMA revival is a confirmed dead end on this recipe — the no-EMA decision from Wave 1 was correct.

**You will be reassigned to a fresh Wave 4 hypothesis shortly. Stand by for your next PR assignment.**

Close this PR as: EMA revival rejected — oscillation unstable; no-EMA Wave 1 decision confirmed.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #181 because EMA revival rejected. Run `scefipy4` showed severe oscillation: 3 regressions across 12 epochs (ep6=15.29%, ep10=12.10%, ep12=12.15%), stuck in 10.8–12.2% band. Ep15 gate (<9.0%) unachievable. EMA=0.999 is too aggressive for this recipe — it tracks noisy gradients rather than smoothing them. The no-EMA decision from Wave 1 is confirmed. Trial B (EMA=0.9995) not launched — the instability is structural. Thorfinn reassigned to Wave 4 hypothesis..
